solutions/usaco-787.mdx

```diff
@@ -87,4 +86,81 @@ int main() {
 ```
 
 </CPPSection>
+<JavaSection>
+```java
+import java.io.BufferedReader;
+import java.io.FileReader;
+import java.io.IOException;
+import java.io.PrintWriter;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.StringTokenizer;
+
+public class rental {
+    public static void main(String[] args) throws IOException {
+        BufferedReader br = new BufferedReader(new FileReader("rental.in"));
+        PrintWriter out = new PrintWriter("rental.out");
+        StringTokenizer s = new StringTokenizer(br.readLine());
+        int n = Integer.parseInt(s.nextToken());
+        int m = Integer.parseInt(s.nextToken());
+        int r = Integer.parseInt(s.nextToken());
+        Integer [] production = new Integer [n];
+        for (int i = 0; i < n; i++) {
+            production[i] = Integer.parseInt(br.readLine());
+        }
+        Integer [][] shops = new Integer [m][2];
+        for (int i = 0; i < m; i++) {
+            s = new StringTokenizer(br.readLine());
+            shops[i][1] = Integer.parseInt(s.nextToken());
+            shops[i][0] = Integer.parseInt(s.nextToken());
+        }
+        Integer [] rent = new Integer [r];
+        for (int i = 0; i < r; i++) {
+            rent[i] = Integer.parseInt(br.readLine());
+        }
+        Arrays.sort(rent, Collections.reverseOrder());
+        Arrays.sort(shops, new Comparator<Integer[]>() {
+            @Override
+            public int compare(Integer[] a, Integer[] b) {
+                return b[0] - a[0];
+            }
+        });
+        Arrays.sort(production, Collections.reverseOrder());
+        int iShop = 0;
+        int iRent = 0;
+        int i = 0;
+        long sol = 0;
+        while (i < n) {
+            int litres = production[i];
+            int moneyT = 0, curI = iShop;
+            int last = 0;
+            while (curI < m) {
+                int rest = Math.min(litres, shops[curI][1]);
+                moneyT += shops[curI][0] * rest;
+                litres -= rest;
+                if (litres == 0) {
+                    last = rest;
+                    break;
+                } else {
+                    curI++;
+                }
+            }
+            if (iRent >= r || moneyT > rent[iRent]) {
+                sol += moneyT;
+                iShop = curI;
+                if (iShop < m) shops[iShop][1] -= last;
+                i++;
+            } else {
+                sol += rent[iRent];
+                n--;
+                iRent++;
+            }
+        }
```

Comment:
Would prefer if you can somewhat comment the core section of your code as the problem involves complex logic. It is easier for the readers to ensure / understand that you covered every possible transition scenario.

Reply:
^^

Reply:
The core component of my code is pretty much directly copied from the existing c++ solution on USACO guide.  Should I add comments to both of those solutions then?

Reply:
Just comment out your own solution should be good.

Reply:
Yes you can just copy over the comments from the c++ sol